src/DependencyInjection/DI.Specification.Tests/src/DependencyInjectionSpecificationTests.cs

```diff
@@ -545,6 +545,160 @@ public void OpenGenericServicesCanBeResolved()
+            var poco = new PocoClass();
+            collection.AddSingleton(poco);
+            collection.AddSingleton<IFakeSingletonService, FakeService>();
+            var provider = CreateServiceProvider(collection);
+            // Act
+            var allServices = provider.GetServices<IFakeOpenGenericService<PocoClass>>().ToList();
+            var constrainedServices = provider.GetServices<IFakeOpenGenericService<IFakeSingletonService>>().ToList();
+            var singletonService = provider.GetService<IFakeSingletonService>();
+            // Assert
+            Assert.Equal(2, allServices.Count);
+            Assert.Same(poco, allServices[0].Value);
+            Assert.Same(poco, allServices[1].Value);
+            Assert.Equal(1, constrainedServices.Count);
+            Assert.Same(singletonService, constrainedServices[0].Value);
+        }
+
+        [Fact]
+        public void ConstrainedOpenGenericServicesReturnsEmptyWithNoMatches()
+        {
+            // Arrange
+            var collection = new TestServiceCollection();
+            collection.AddTransient(typeof(IFakeOpenGenericService<>), typeof(ConstrainedFakeOpenGenericService<>));
+            collection.AddSingleton<IFakeSingletonService, FakeService>();
+            var provider = CreateServiceProvider(collection);
+            // Act
+            var constrainedServices = provider.GetServices<IFakeOpenGenericService<IFakeSingletonService>>().ToList();
+            // Assert
+            Assert.Equal(0, constrainedServices.Count);
+        }
+
+        [Fact]
+        public void InterfaceConstrainedOpenGenericServicesCanBeResolved()
+        {
+            // Arrange
+            var collection = new TestServiceCollection();
+            collection.AddTransient(typeof(IFakeOpenGenericService<>), typeof(FakeOpenGenericService<>));
+            collection.AddTransient(typeof(IFakeOpenGenericService<>), typeof(ClassWithInterfaceConstraint<>));
+            var enumerableVal = new ClassImplementingIEnumerable();
+            collection.AddSingleton(enumerableVal);
+            collection.AddSingleton<IFakeSingletonService, FakeService>();
+            var provider = CreateServiceProvider(collection);
+            // Act
+            var allServices = provider.GetServices<IFakeOpenGenericService<ClassImplementingIEnumerable>>().ToList();
+            var constrainedServices = provider.GetServices<IFakeOpenGenericService<IFakeSingletonService>>().ToList();
+            var singletonService = provider.GetService<IFakeSingletonService>();
+            // Assert
+            Assert.Equal(2, allServices.Count);
+            Assert.Same(enumerableVal, allServices[0].Value);
+            Assert.Same(enumerableVal, allServices[1].Value);
+            Assert.Equal(1, constrainedServices.Count);
+            Assert.Same(singletonService, constrainedServices[0].Value);
+        }
+
+        [Fact]
+        public void PublicNoArgCtorConstrainedOpenGenericServicesCanBeResolved()
```

Comment:
I think this test has an invalid premise, so I'm hoping to get clarification on the intent. I think it may have an impact on how the spec for constraints should work in a single instance case.

Here's how I'm reading it:

- Register an open generic with no constraints. This should work for any `T` in the generic.
- Register an open generic with a constraint of `new()`. This will only work for things with a public constructor.
- Resolve all the services using a class that can be constructed. You should get two.
- Resolve all the services using a class that doesn't satisfy the constraint. The constraint violation will be ignored but the others will be resolved.

A challenge I'm having is that this impacts the way a single instance resolution would work.

```csharp
var collection = new TestServiceCollection();
collection.AddTransient(typeof(IFakeOpenGenericService<>), typeof(ClassWithNoConstraints<>));
collection.AddTransient(typeof(IFakeOpenGenericService<>), typeof(ClassWithNewConstraint<>));
var provider = CreateServiceProvider(collection);
// What do you get here? Exception? Fallback to ClassWithNoConstraint<T>? null?
var singleService = provider.GetService<IFakeOpenGenericService<ClassWithPrivateCtor>>();
```

From autofac/Autofac.Extensions.DependencyInjection#40 I gather the intent is...

- In the single instance, you'd get an exception. The last thing you registered - the override - doesn't satisfy the constraint, so resolution fails.
- In the collection, you get a sort of "filtered" collection. Only things that satisfy the constraint are resolved.

From a use case perspective, let's say I have a bunch of message handlers. I register them all and resolve the collection of them. Sometimes I get them all and sometimes I don't. How do I know why? How do I troubleshoot that? 

I'm not sure I agree with the filtered collection concept. It may pass on the base container, but I'm not sure it should. I'd expect any constraint violation to notify me. Perhaps I'm alone in that?

Reply:
Don't think of it as "resolving a collection". I'm not registering a static collection. I'm registering a bunch of individual handlers, then I come back later and ask "find all the handlers that match this specification".